# Preprocessing


In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID columns, 'EIN'.
application_df = application_df.drop(columns=['EIN'])
application_df.head()

NAME APPLICATION_TYPE  \
0              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [4]:
# Determine the number of unique values in each column.
application_df.nunique()

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [5]:
# Look at NAME value counts for binning
application_name_counts = application_df['NAME'].value_counts()
application_name_counts

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [6]:
# Choose a cutoff value and create a list of names_replace to be replaced
# use the variable name names_to_replace

cutoff_value = 5
names_to_replace = application_name_counts[application_name_counts < cutoff_value].index.tolist()

# Replace names in the dataframe with 'Other'
for name in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(name, 'Other')

# Check to make sure binning was successful
application_df['NAME'].value_counts()

Other                                                                19803
PARENT BOOSTER USA INC                                                1260
TOPS CLUB INC                                                          765
UNITED STATES BOWLING CONGRESS INC                                     700
WASHINGTON STATE UNIVERSITY                                            492
                                                                     ...  
GUILDS OF THE SANTA FE OPERA INC                                         5
FAMILY MOTOR COACH ASSOCIATION                                           5
JAMESTOWN SOCIETY                                                        5
MOST WORSHIPFUL PRINCE HALL GRAND LODGE OF FREE & ACCEPTED MASONS        5
PTA HAWAII CONGRESS                                                      5
Name: NAME, Length: 403, dtype: int64

In [7]:
# Look at APPLICATION_TYPE value counts for binning
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
application_type_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff_value = 500
application_types_to_replace = application_type_counts[application_type_counts < cutoff_value].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app, "Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [9]:
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts
classification_counts

<bound method IndexOpsMixin.value_counts of 0        C1000
1        C2000
2        C3000
3        C2000
4        C1000
         ...  
34294    C1000
34295    C3000
34296    C2000
34297    C3000
34298    C1000
Name: CLASSIFICATION, Length: 34299, dtype: object>

In [10]:
# Look at CLASSIFICATION value counts where the count is greater than 1
classification_counts = application_df["CLASSIFICATION"].value_counts()
classification_counts_gt1 = classification_counts[classification_counts > 1]

classification_counts_gt1

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [11]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff_value = 1000
classifications_to_replace = classification_counts[classification_counts < cutoff_value].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df,dtype=float)
application_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_AACE INTERNATIONAL  \
0       1     5000              1                      0.0   
1       1   108590              1                      0.0   
2       1     5000              0                      0.0   
3       1     6692              1                      0.0   
4       1   142590              1                      0.0   

   NAME_ACE MENTOR PROGRAM OF AMERICA INC  NAME_ACTS MINISTRY  \
0                                     0.0                 0.0   
1                                     0.0                 0.0   
2                                     0.0                 0.0   
3                                     0.0                 0.0   
4                                     0.0                 0.0   

   NAME_ACTS MISSIONS  \
0                 0.0   
1                 0.0   
2                 0.0   
3                 0.0   
4                 0.0   

   NAME_AFRICAN-AMERICAN POSTAL LEAGUE UNITED FOR SUCCESS A-PLUS  \
0                                                0.0               
1                                                0.0               
2                                                0.0               
3                                                0.0               
4                                                0.0               

   NAME_AIR FORCE ASSOCIATION  NAME_ALABAMA FEDERATION OF WOMENS CLUBS  ...  \
0                         0.0                                      0.0  ...   
1                         0.0                                      0.0  ...   
2                         0.0                                      0.0  ...   
3                         0.0                                      0.0  ...   
4                         0.0                                      0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 447 columns]

In [13]:
# Split our preprocessed data into our features and target arrays

X = application_df.drop(columns="IS_SUCCESSFUL")

y = application_df["IS_SUCCESSFUL"]

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Compile, Train and Evaluate the Model

In [15]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.


# Create a sequential model
nn = tf.keras.models.Sequential()

# Input layer
nn.add(tf.keras.layers.Dense(units=128, activation='relu', input_dim=X_train_scaled.shape[1]))


# First hidden layer
nn.add(tf.keras.layers.Dense(units=64, activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=32, activation='relu'))


# Third hidden layer
nn.add(tf.keras.layers.Dense(units=16, activation='relu'))


# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))


# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 128)               57216     
                                                                 
 dense_1 (Dense)             (None, 64)                8256      
                                                                 
 dense_2 (Dense)             (None, 32)                2080      
                                                                 
 dense_3 (Dense)             (None, 16)                528       
                                                                 
 dense_4 (Dense)             (None, 1)                 17        
                                                                 
Total params: 68097 (266.00 KB)
Trainable params: 68097 (266.00 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [16]:
# Compile the model
nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [17]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,validation_split=0.15, epochs=100)

Epoch 1/100
729/729 [==============================] - 4s 4ms/step - loss: 0.4700 - accuracy: 0.7788 - val_loss: 0.4507 - val_accuracy: 0.7869
Epoch 2/100
729/729 [==============================] - 3s 4ms/step - loss: 0.4272 - accuracy: 0.7970 - val_loss: 0.4423 - val_accuracy: 0.7923
Epoch 3/100
729/729 [==============================] - 2s 3ms/step - loss: 0.4228 - accuracy: 0.7983 - val_loss: 0.4450 - val_accuracy: 0.7937
Epoch 4/100
729/729 [==============================] - 2s 3ms/step - loss: 0.4206 - accuracy: 0.7992 - val_loss: 0.4458 - val_accuracy: 0.7918
Epoch 5/100
729/729 [==============================] - 2s 3ms/step - loss: 0.4190 - accuracy: 0.8006 - val_loss: 0.4456 - val_accuracy: 0.7898
Epoch 6/100
729/729 [==============================] - 3s 4ms/step - loss: 0.4176 - accuracy: 0.8007 - val_loss: 0.4432 - val_accuracy: 0.7913
Epoch 7/100
729/729 [==============================] - 3s 3ms/step - loss: 0.4174 - accuracy: 0.8020 - val_loss: 0.4450 - val_accuracy: 0.7908

In [18]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 1s - loss: 0.5139 - accuracy: 0.7923 - 665ms/epoch - 3ms/step
Loss: 0.513899564743042, Accuracy: 0.7922740578651428


In [19]:
# Save and export the results to an HDF5 file
# Export model to HDF5 File

from google.colab import files

nn.save('/content/AlphabetSoupCharity_Optimization.h5')
files.download('/content/AlphabetSoupCharity_Optimization.h5')

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Alphabet Soup Funding Analysis Report
**Overview of the Analysis:**
The primary objective of this analysis is to leverage the power of machine learning, particularly deep learning, to create a binary classification model that can predict the success of applicants funded by Alphabet Soup. This tool will assist Alphabet Soup in making informed decisions about which organizations to fund based on the likelihood of their success.

**Results:**

**Data Preprocessing:**

**Target Variable:**

**IS_SUCCESSFUL:** This indicates whether the funds provided by Alphabet Soup were used effectively by the organization.

**Features:**

All the columns present in the dataset, except for EIN and IS_SUCCESSFUL, were used as features. This includes APPLICATION_TYPE, AFFILIATION, CLASSIFICATION, USE_CASE, ORGANIZATION, STATUS, INCOME_AMT, SPECIAL_CONSIDERATIONS, ASK_AMT, and NAME.

**Variables Removed:**

**EIN:** This column was removed as it is an identification column and does not provide any significant information to determine the success of an organization.

**Compiling, Training, and Evaluating the Model:**

**Model Structure:**

Input Layer: 128 neurons with ReLU activation function.

First Hidden Layer: 64 neurons with ReLU activation function.

Second Hidden Layer: 32 neurons with ReLU activation function.

Third Hidden Layer: 16 neurons with ReLU activation function.

Output Layer: 1 neuron with Sigmoid activation function since it's a binary classification problem.

**Model Compilation:**

Optimizer: Adam

Loss function: Binary Crossentropy

Metrics: Accuracy

**Model Performance:**

The model's accuracy on the test set was found to be 79%.

**Steps to Optimize Model Performance:**

The data was preprocessed by binning values of categorical columns that had a large number of unique values. Specifically, values in NAME, APPLICATION_TYPE, and CLASSIFICATION columns were grouped into an 'Other' category if they occurred fewer times than a set cutoff value.
Data was then converted into a numerical format using one-hot encoding (pd.get_dummies).

The data was standardized using StandardScaler to have zero mean and unit variance.

The deep learning model was designed with multiple layers and neurons to capture the underlying patterns in the data. The model was trained over 100 epochs.

**Summary:**

The deep learning model designed for predicting the success of applicants funded by Alphabet Soup provides a mechanism for understanding the key factors that determine whether an organization will use the funds effectively. The model's accuracy indicates its effectiveness in making these predictions.

However, deep learning models can sometimes be seen as a 'black box', and it might be challenging to interpret which features had the most influence on the predictions.

**Recommendation:**

A different approach could be to use ensemble learning methods, such as Random Forest. This model can provide insights into feature importance, allowing Alphabet Soup to understand which factors are the most critical in determining an organization's success. This is also less prone to overfitting compared to deep neural networks and can sometimes achieve comparable or even better performance with the right hyperparameter tuning.